### ISSUE TITLE
Support creation of type aliases for delegate types via Imports statement.

### ISSUE THREAD

**Issue by ericmutta** [dotnet/vblang#350]
Discussion related to [#30724](https://github.com/dotnet/roslyn/issues/30724). Basically, this code should compile:

```vb
Imports MyFunction = System.Func(Of String) '<-- compiler doesn't like this.

Public Module Module1
  Public Sub Main()

  End Sub
End Module
```

...because the language spec [says that it should](https://docs.microsoft.com/en-us/dotnet/visual-basic/reference/language-specification/source-files-and-namespaces#import-aliases):

> An import alias can create an alias **for any namespace or type**. Accessing a namespace or type through an alias yields exactly the same result as accessing the namespace or type through its declared name

Creating type aliases is very useful when dealing with complex generic types (e.g a dictionary where the values are themselves dictionaries) so removing the above limitation would be a small but useful enhancement.

While on the topic, it would also be great to remove the limitation that says you can't refer to other type aliases when defining new aliases. For example this doesn't compile:

```vb
Imports StringList = System.Collections.Generic.List(Of String)
Imports StringListTable = System.Collections.Generic.Dictionary(Of String, StringList)
```

...because you can't refer to the `StringList` alias when defining the `StringListTable`  alias. Combined with the rule that the types must be fully qualified, you end up having to type this keyboard killer:

```vb
Imports StringListTable = System.Collections.Generic.Dictionary(Of String, System.Collections.Generic.List(Of String))
```

In summary, type aliases can be made a great deal more useful by making the following changes:

1. Allow creation of type aliases for delegate types like `System.Func(Of T)`.
2. Allow creation of type aliases without using fully qualified type names when possible.
3. Allow creation of type aliases that refer to other type aliases defined in same file.
4. Allow creation of type aliases in scopes smaller than the file (e.g inside a single class/method).

Number 1 is the primary request in this issue, but if the language team is willing to look at the area as a whole, then features number 2 to 4 above would be great to include under the umbrella of **improved alias types** in a future version of VB.

CC: @KathleenDollard

**Comment by rskar-git** [comment]
> [1]. Allow creation of type aliases for delegate types like System.Func(Of T).
> [3]. Allow creation of type aliases that refer to other type aliases defined in same file.

I very much like 1. and 3. and hope these can be done sometime soon!

> [4]. Allow creation of type aliases in scopes smaller than the file (e.g inside a single class/method).

Item 4. is a nice to have, I suppose, but it doesn't seem all that essential. I think I  would prefer to have the aliases (for types) spelled out toward the top anyway, rather than having a potential "free-for-all" scattered about the source code.

> [2]. Allow creation of type aliases without using fully qualified type names when possible.

This is the one item I would down-vote. The `Imports` is an important bit of documentation that I think should unambiguously specify exactly which type is being aliased. So I think a fully qualified type name should be used every time. If 3. becomes a reality, then fully qualified names won't be onerous at all.

**Comment by ericmutta** [comment]
> @rskar-git: Item 4. is a nice to have, I suppose, but it doesn't seem all that essential.

I forgot to specify the rationale behind it. For example this DOES compile:

```vb
Imports StringFoo = Project1.Outer.Foo(Of String)

Public Class Outer
  Public Class Foo(Of T) '<-- class Foo is public.
  End Class
End Class
```

...but the following does NOT compile:

```vb
Imports StringFoo = Project1.Outer.Foo(Of String)

Public Class Outer
  Private Class Foo(Of T) '<-- class Foo is private.
  End Class
End Class
```

The compiler complains that `Foo` is private and is therefore not accessible when defining the alias `StringFoo`. I suppose this is fine for preventing access to class Foo in other parts of the file outside class `Outer` but it also means you can't define an alias for `Foo(Of T)` for use exclusively inside class `Outer`. This is why I suggested allowing alias types to be defined in scopes smaller than "the whole file".

**The alternative to suggestion 4 would be to show the compiler error on REFERENCES to the alias `StringFoo` rather than on the DEFINITION (i.e you can use `StringFoo` anywhere inside class `Outer`, but not outside it where class `Foo(Of T)` is not visible. In essence, the alias type inherits the scope of the type being aliased).**

> @rskar-git:  If 3. becomes a reality, then fully qualified names won't be onerous at all.

This is an excellent observation and allows suggestion number 2 to be removed completely :+1: 

So for convenient reference based on all the above, here is a revised set of suggestions for improving alias types:

1. Allow creation of type aliases for delegate types like `System.Func(Of T)`.
2. Allow creation of type aliases that refer to other type aliases defined in same file.
3. Allow creation of type aliases for private types and restrict usage to the private type's scope.
